File: Go/Commands/Mkdisk.go
package Commands

import (
	"bytes"
	"encoding/binary"
	"fmt"
	"log"
	"os"
	"strconv"
	"time"
)

func Create_Disk(unit string, path string, fit string, size string) {
	size_, err := strconv.Atoi(size)
	if err != nil {
		log.Fatal(err)
	}
	if path == "" {
		fmt.Println(red + "Error: Falta el parametro obligatorio Path, intentelo denuevo." + closing)
	} else if size_ <= 0 {
		fmt.Println(red + "Error: El Size no puede ser 0 o menor a 0, cambielo e intentelo denuevo." + closing)
	} else if FileExist(path) == true {
		fmt.Println(red + "Error: El disco " + yellow + find_last_of(path) + closing + red + " ya existe." + closing)
	} else {
		dt := time.Now()
		size_identified := new_size(unit, size_)
		ceros_ := make([]byte, size_identified)
		for i := 0; i < size_identified; i++ {
			ceros_[i] = 0
		}
		file, err := os.Create(path)
		if err != nil {
			fmt.Println(red + "Ocurrio un error inesperado, intente denuevo." + closing)
		}
		file.Write(ceros_)
		signature := DiskSignature()
		var bufferControl bytes.Buffer
		mbr_object := MBR{}
		copy(mbr_object.mbr_path[:], path)
		copy(mbr_object.mbr_date[:], dt.String())
		copy(mbr_object.mbr_size[:], strconv.Itoa(size_identified))
		copy(mbr_object.mbr_disk_signature[:], strconv.Itoa(signature))
		copy(mbr_object.mbr_fit[:], fit)
		for i := 0; i < 4; i++ {
			copy(mbr_object.mbr_particion[i].part_status[:], "0")
			copy(mbr_object.mbr_particion[i].part_type[:], "P")
			copy(mbr_object.mbr_particion[i].part_fit[:], fit)
			copy(mbr_object.mbr_particion[i].part_start[:], "0")
			copy(mbr_object.mbr_particion[i].part_size[:], "0")
			copy(mbr_object.mbr_particion[i].part_name[:], "")
		}
		file.Seek(0, os.SEEK_SET)
		mbr_in_bytes := Convert_to_Bytes(mbr_object)
		binary.Write(&bufferControl, binary.BigEndian, &mbr_in_bytes)
		WriteBytes(file, bufferControl.Bytes())
		file.Close()
		Print_Msg_Disk_Created(dt, path, size_identified, signature)
	}
}

File: C++/Commands/mkdisk.cpp
#include <iostream>
#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <string.h>
#include <cstdlib>
#include <time.h>
#include "estructures.h"
using namespace std;


struct Particion
{
    char part_status;
    char part_type = 'p'; 
    char part_fit = 'f'; 
    int part_start;
    int part_size;
    char part_name[16];
};
struct MBR
{
    int mbr_tamano;
    time_t mbr_fecha_creacion;
    int mbr_disk_signature;
    string disk_fit;
    Particion mbr_partition[4];
};

int new_size(int size_, string unit){
    if (unit == "b"){
        return size_;
    }else if (unit == "k"){
        return size_ * 1024;
    }else if (unit == "m"){
        return size_ * 1048576;
    }
    return 0;
}



void CreateDisc(string path, string size_, string unit, string fit){
    string s = path;
    char sc[s.size() + 1];
    strcpy(sc, s.c_str());
    FILE *file=NULL;
    file=fopen(sc,"r");
    if(file!=NULL){
        cout<<"Ya existe el disco"<<endl;
        return;
    }
    int size_converted = stoi(size_);
    int tam = (new_size(size_converted, unit));;

    file=fopen(sc,"wb");
    fwrite("\0",1,1,file);
    fseek(file,tam,SEEK_SET); 
    fwrite("\0",1,1,file);
    MBR mbr;
    mbr.mbr_tamano = tam;
    mbr.mbr_disk_signature = rand()%1000;
    mbr.mbr_fecha_creacion = time(0);
    mbr.disk_fit = fit;
    for(int i = 0; i < 4; i++){
        mbr.mbr_partition[i].part_status = '0';
        mbr.mbr_partition[i].part_size = 0;
        mbr.mbr_partition[i].part_fit = 'f';
        mbr.mbr_partition[i].part_start = tam;
        strcpy(mbr.mbr_partition[i].part_name,"");
    }
    cout<<"-------------DISCO CREADO--------------------"<<endl;
    cout<<"Fecha de creacion: "<<asctime(gmtime(&mbr.mbr_fecha_creacion))<<endl;
    cout<<"Signature: "<<mbr.mbr_disk_signature <<endl;
    cout<<"Tamanio: "<<mbr.mbr_tamano <<endl;
    cout<<"Fit: " <<mbr.disk_fit <<endl;
    fseek(file,0,SEEK_SET);
    fwrite(&mbr,sizeof(MBR),1,file);
    fclose(file);
}